athena-elasticsearch/src/main/java/com/amazonaws/connectors/athena/elasticsearch/ElasticsearchMetadataHandler.java

```diff
@@ -0,0 +1,282 @@
+        this.awsGlue = getAwsGlue();
+        this.clientFactory = ElasticsearchHelper.getClientFactory();
+
+        ElasticsearchHelper.setDomainMapping();
+    }
+
+    @VisibleForTesting
+    protected ElasticsearchMetadataHandler(AWSGlue awsGlue,
+                                           EncryptionKeyFactory keyFactory,
+                                           AWSSecretsManager awsSecretsManager,
+                                           AmazonAthena athena,
+                                           String spillBucket,
+                                           String spillPrefix,
+                                           AwsRestHighLevelClientFactory clientFactory)
+    {
+        super(awsGlue, keyFactory, awsSecretsManager, athena, SOURCE_TYPE, spillBucket, spillPrefix);
+        this.awsGlue = awsGlue;
+        this.clientFactory = clientFactory;
+
+        ElasticsearchHelper.setDomainMapping();
+    }
+
+    /**
+     * Used to get the list of domains (aka databases) for the Elasticsearch service.
+     * @param allocator Tool for creating and managing Apache Arrow Blocks.
+     * @param request Provides details on who made the request and which Athena catalog they are querying.
+     * @return A ListSchemasResponse which primarily contains a Set<String> of schema names and a catalog name
+     * corresponding the Athena catalog that was queried.
+     */
+    @Override
+    public ListSchemasResponse doListSchemaNames(BlockAllocator allocator, ListSchemasRequest request)
+    {
+        logger.info("doListSchemaNames: enter - " + request);
+
+        if (ElasticsearchHelper.isAutoDiscoverEndpoint()) {
+            // Since adding/deleting domains in Amazon Elasticsearch Service doesn't re-initialize
+            // the connector, the domainMap needs to be refreshed each time for Amazon ES.
+            ElasticsearchHelper.setDomainMapping();
+        }
+
+        return new ListSchemasResponse(request.getCatalogName(), ElasticsearchHelper.getDomainList());
+    }
+
+    /**
+      * Used to get the list of indices contained in the specified domain.
+      * @param allocator Tool for creating and managing Apache Arrow Blocks.
+      * @param request Provides details on who made the request and which Athena catalog and database they are querying.
+      * @return A ListTablesResponse which primarily contains a List<TableName> enumerating the tables in this
+      * catalog, database tuple. It also contains the catalog name corresponding the Athena catalog that was queried.
+      */
+    @Override
+    public ListTablesResponse doListTables(BlockAllocator allocator, ListTablesRequest request)
+    {
+        logger.info("doListTables: enter - " + request);
+
+        List<TableName> indices = new ArrayList<>();
+        String endpoint = ElasticsearchHelper.getDomainEndpoint(request.getSchemaName());
+
+        if (!endpoint.isEmpty()) {
+            AwsRestHighLevelClient client = clientFactory.getClient(endpoint);
```

Comment:
Do we need to create a new instance of the client every time or can we have a client cache so that we reuse the client whenever possible?

Reply:
The client requires an endpoint which is injected as part of the creation. Since the implementation supports multiple endpoints, it makes sense to instantiate the client right before it is used. Caching or pooling the clients for each endpoint would imply either having another static member with static accessors, or duplication as each handler would need its own cache of clients. Furthermore, each instantiated client needs to be properly closed which makes its instantiation more intuitive within the confines of a try/catch/finally clause. From the Elasticsearch docs: "It is important that it gets closed when no longer needed so that all the resources used by it get properly released, as well as the underlying http client instance and its threads."

Reply:
re-opened this conversation

Reply:
Fine with postponing this change until multi-split implementation